File: Data Structures & Algorithms/top-k-elements-in-list/submission-6.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        int maxFreq = 0;
        Map<Integer, Integer> freqMap = new HashMap<>();
        for(int num : nums){
            int freq = freqMap.getOrDefault(num,0)+1;
            freqMap.put(num, freq);
            maxFreq = Math.max(maxFreq, freq);
        }
        // To sort the values and bucket it -> I know a technique called bucket sort
        List<Integer>[] buckets = new ArrayList[maxFreq+1];
        for(Map.Entry<Integer, Integer> entry : freqMap.entrySet()){
            Integer value = entry.getKey();
            Integer freq = entry.getValue();
            if(buckets[freq] == null){
                buckets[freq] = new ArrayList();
            }
            buckets[freq].add(value);
        }

        List<Integer> result = new ArrayList<>();
        int i = buckets.length - 1;
        while(i > 0 && result.size() < k){
            if(buckets[i]!= null){
                for(int val: buckets[i]){
                    result.add(val);
                }
            }
            i--;
        }

        return result.stream().mapToInt(Integer::intValue).toArray();


    }
}

File: Data Structures & Algorithms/house-robber/submission-6.js
class Solution {

    map = {}

    findMax(nums,n){

        if(this.map[n]!== undefined) return this.map[n];

        if( n === 0){
            return nums[0];
        }

        if(n === 1){
            return Math.max(nums[1], nums[0]);
        }

        if(n<0){
            return 0;
        }

        const maxRob =  Math.max(nums[n]+this.findMax(nums,n-2), this.findMax(nums, n-1));
        this.map[n] = maxRob;
        return maxRob;
    }

    /**
     * @param {number[]} nums
     * @return {number}
     */
    rob(nums) {
        const n = nums.length - 1;
        return this.findMax(nums,n)
    }
}
